### DISCUSSION TITLE
{{newline}} variable (was: Support linebreaks by using \n, \r and \r\n)

### DISCUSSION THREAD

**Discussion by Taitava** [Taitava/obsidian-shellcommands#42] [Ideas - on hold]
From #24:

> These ‘modes’ could be useful not only for multiple file_paths, but for text in multiple lines too
> - separator could be a \n Line break, adding one line break after each line from clipboard

`\n` is a must have feature at some point, but it's too bad that it doesn't work right now. Currently e.g. `echo \n >> SomeNote.md` writes literally *\n* to a file, at least on Windows. I need to think about how to implement it, can I just replace all `\n`s in the command text with a newline? Probably not, as there is this mess: Linux and Mac use `\n` (and classic Mac used `\r`), but Windows uses `\r\n`. So maybe I should replace `\n` with a linefeed character and `\r` with a carriage return character ([more about these characters on Stack Overflow](https://stackoverflow.com/a/3098328/2754026)). I feel a bit uncertain about being able to implement this correctly, so this won't be coming soon.

Another thing that might work, would be a new variable: `{{br:n}}`. This would create linebreaks (*br* comes from HTML's [`<br>` tag](https://www.w3schools.com/tags/tag_br.asp), but I'm not certain if I'm happy with the *br* name or not). `n` would be a number to indicate how many linebreaks you want, so that you don't have to write anything stupid like `{{br}}{{br}}{{br}}` for three linebreaks, when `{{br:3}}` does the same thing. The argument could be left out: `{{br}}` would give you just a single linebreak. The *n* argument would be a long term reason for the existence of this variable: Even if `\n` (etc) will be supported in the future, `{{br}}` with it's repeator argument would still be useful. And `{{br}}` would adapt to use a correct linebreak in each operating system.

**Comment by FelipeRearden** [comment]
Ooowww, this new {{variable:argument}} idea is fantastic @Taitava !!!!!

Could bring text transportation on Shell Commands to a new level!

**Comment by Taitava** [comment]
Now I think `\n`/`\r`/`\r\n` markings should not be converted by SC into newlines. It might be unexpected to a user. In Mac `\n` seems to already translate into a newline character, but in Windows it does not. I feel I should not add any other subtitutions to commands in addition to the `{{` `}}` syntax.

As for the second idea about a `{{variable}}` for a newline character, I've decided to implement it. The name will be `{{newline}}` instead of `{{br}}`, because I do not like abbreviations in variable names (not everybody knows what they mean), and *br* reminds me too much about HTML's `<br>` tag, and I have to point out that this variable will not output *\<br>*, it will output:
 - `\n` on Mac and Linux.
 - ~~`^` + carriage return + line feed on Windows.~~ Edit: This only works in normal terminal, not in SC.

**Reply by FelipeRearden** [reply]
> In Mac \n seems to already translate into a newline character, but in Windows it does not. 

Just to confirm that `\n` works on Mac.

> The name will be {{newline}} instead of {{br}}, because I do not like abbreviations in variable names (not everybody knows what they mean), and br reminds me too much about HTML's <br> tag, and I have to point out that this variable will not output <br>, it will output:

Great idea @Taitava !!!!

**Comment by Taitava** [comment]
I'll put this on hold, because the value I was going to use for Windows CMD did not work.

**Reply by Taitava** [reply]
I'm now implementing this so that if CMD is used, the variable simply does not work and will give an error message. Other shells should work ok - at least I hope so.

**Reply by Taitava** [reply]
I might change my mind again. As there would not be CMD support for this variable anyway, I think there is not so much use for this variable:
- `\n` works on Bash (used on macOS and Linux) - so they don't really need a variable for this.
- For Windows, PowerShell supports `` `n `` as a newline character.

I have already implemented this variable (it's just not released yet), but when thinking about it more, I feel that I do not want to grow the list of variables with something that is not beneficial. I'll probably make a commit that reverts this new variable, so removes the variable, but the code for the variable stays in Git, so it can be later revived if wanted.

@FelipeRearden, what do you think, do you see any use for this variable?

**Reply by FelipeRearden** [reply]
> @FelipeRearden, what do you think, do you see any use for this variable?

I agree with you @Taitava today I don't see any diference between {{newline}} and \n . Since we can use \n in every way possible. 

The only thing that could be useful is {{newline:separator}} but I don't have in my mind (or in my FR pocket list) today any workflow for the separator. 

I will keep this in my mind and I let you now if I figure something out. 

But I agree that we should hold it for now :)

**Comment by Taitava** [comment]
# Implemented as a debug variable

Just to note this here too, I created the `{{newline}}` variable, but it returns `\n` regardless of operating system. Can be changed later, if needed. However, the variable is only available in debug mode, and meant for testing how SC manipulates newline characters that come from variable values.

For real world usage, you can add newline characters by pressing enter in the shell command field.

#295
